book/src/developers/designs/domain_join.md

```diff
@@ -0,0 +1,111 @@
+
+In this case when a user authenticates with ssh, since we are a trusted machine we should be able to
+request short-term and limited credentials on the users behalf.
+
+An example is that we could dynamicly request TLS certificates or kerberos credentials.
+
+Normally with ssh in this manner, everything has to use kerberos. This would force users to kinit on
+their machine to ssh and forward their credentials to the next machine. This causes friction since
+configuring kerberos on machines is an exercise in frustration, and with BYOD it gets even worse. In
+addition if a use ssh's with an ssh key, the only viable kinit mechanism is password or password + totp. Both
+of which are falling out of favour for cryptographic authentication.
+
+## Security Considerations
+
+* Anonymous joins should not be allowed or permitted.
+* Join tokens need to be revoked (causing related machines to request re-enrollment) or expired (related machines can continue to function)
+* Join tokens must be auditable.
+* Private keys SHOULD be stored in a TPM, or at least a software HSM with a secured unlock key.
+* The use of the private key must prevent replay attacks
+
+## Overview
+
+Since the machine would now be an entity requiring authentication, we need to have a process to
+establish and maintain this trust relationship.
+
+1. A join token is created by a user who is authorised to perform domain joins.
+2. The machine is audited for a known trust state. This process may vary from site to site. A future improvement could be that the join token can only release on certain TPM PCR values.
+3. The join token is yielded to the kanidm unix daemon which submits it's signing key to the kanidm server.
+4. The kanidm server verifies the submission and creates a machine account.
+5. The kanidm unix daemon now uses it's signing key to sign a challenge that is submitted with all requests to the kanidm server.
+
+Extra
+
+6. Machines should be able to "re-join" with an alternate join token, moving their machine account join token relationship.
+7. Machines must be able to self-enroll newer keys which may have stronger cryptographic requirements.
+
+## Details
+
+### Join Token Creation
+
+Join tokens are persisted in the database allowing tracing back to the usage of the token.
+
+Every machine that is joined by that token will related back to that token. This allows auditing
+of which token was used to join which machine.
+
+Machines may re-enroll with an alternate token.
+
+The join token should be signed. The JWK pub key should be available at a known HTTPS uri so that the client
+can use it to validate the join token and it's content. This *may* allow policy to be embedded into
+the join token for the client to self-adhere to in the join process.
+
+### Machine Auditing
+
+The machine should be audited to be in a secure state. It's not yet clear how to proceed here, but
+we should consider using TPM PCRs with secure boot to measure this and validate the machine state.
```

Comment:
Azure Intune clients for example would check here if the disk is encrypted.

Reply:
Yeah. We need to see how that's working in ALP and other distros generally. Not every distro does FDE with TPM's, meaning that FDE != passing secure boot on all platforms. ALP however is TPM PCR secure boot based, so we can use that as a measure of "some" boot health. I'm also not sure how to programmatically check that either.